modules/shared/src/main/scala/org/tessellation/currency/dataApplication/DataApplicationTraverse.scala

```diff
@@ -211,16 +211,29 @@ object DataApplicationTraverse {
         globalSnapshot.stateChannelSnapshots
           .get(identifier) match {
           case Some(snapshots) =>
-            HasherSelector[F].forOrdinal(globalSnapshot.ordinal) { implicit hasher =>
-              snapshots.toList.traverse { binary =>
-                jsonBrotliBinarySerializer.deserialize[Signed[CurrencyIncrementalSnapshot]](binary.content)
-              }
-                .map(_.flatMap(_.toOption))
-                .map(NonEmptyList.fromList)
-                .map(_.map(_.sortBy(_.value.ordinal)))
-                .flatMap(_.map(_.traverse(_.toHashedWithSignatureCheck)).sequence)
-                .map(_.map(_.traverse(_.toValidatedNel)))
+            snapshots.toList.traverse { binary =>
+              jsonBrotliBinarySerializer.deserialize[Signed[CurrencyIncrementalSnapshot]](binary.content)
             }
+              .map(_.flatMap(_.toOption))
+              .map(NonEmptyList.fromList)
+              .map(_.map(_.sortBy(_.value.ordinal)))
+              .flatMap(_.map(_.traverse { snapshot =>
+                def usingKryo = {
+                  implicit val hasher = Hasher.forKryo[F]
+                  snapshot.toHashedWithSignatureCheck
+                }
+
+                def usingJson = {
+                  implicit val hasher = Hasher.forJson[F]
+                  snapshot.toHashedWithSignatureCheck
+                }
+
+                usingJson.flatMap {
+                  case Left(_)  => usingKryo
+                  case Right(s) => s.asRight[Signed.InvalidSignatureForHash[CurrencyIncrementalSnapshot]].pure[F]
+                }
+              }).sequence)
+              .map(_.map(_.traverse(_.toValidatedNel)))
```

Comment:
This code is duplicated. Is it possible to pull it out into common place?
Or maybe generalize it using `EitherT.valueOrF` ?